src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -134,6 +140,9 @@ public E borrowItem() throws InterruptedException {
             do {
                 if (isPoolLockHeldByAnotherThread(currentThread)) {
                     poolNotLocked.await();
+                } else if (this.poisonPill != null){
+                    // Return the pill immediately if there is one
+                    item = poisonPill;
```

Comment:
I think that it would be ok to return it even if the pool is locked, since receiving the pill indicates the pool is borked and will never be unborked, and since the pill is not a limited resource like the queue that has to be carefully handed out, and since getting the pill doesn't allow the consumer to cause any side effects in the pool anyways. 

I'm not even sure we need to lock the queue before returning the pill, but it seems good to keep all the logic inside the `try/finally` block